src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+
+### Provider Types
+
+There are three different provider types that can be used.
+These three provider along with their `When` and `Causing` DSL words determine the flow of a provider, executing code as the director changes which tasks it executes.
+
+#### Entering
+
+An entering provider is used to set up the state of a provider when it is running, or to ensure that a When condition of a main provider is met.
+They are executed in two scenarios:
+
+1. when a provider group has no active tasks and then is given to one.
+2. when it was running a task normally but the only `Provides` reaction that is running has its when condition fail and needs to re-enter to provide a `Causing`.
+
+Example:
+
+```cpp
+on<Entering<Walk>>().then()
+```
+
+#### Provides
+
+Provides are the normal providers which will be expected to execute a task.
+They will be chosen as the active provider in a group if there are no Entering statements in a provider group, or if at least one entering provider has executed making this group "entered".
+However an entering provider may be repeatedly run still if all the providers have `When` conditions that are not yet met.
+
+Example:
+
+```cpp
+on<Provides<Walk>>().then()
+```
+
+#### Leaving
+
+Leaving providers are used for the state when a provider group is losing control and will no longer have access to subtasks, or when a higher priority task elsewhere needs a causing that it provides.
+Leaving providers are executed in two scenarios:
+
+Firstly, A leaving provider will be executed once when the provider no longer has an active task.
+Note: If there no leaving providers that are satisfied in this situation (due to when conditions) the system will stop executing the provider without informing it.
+It may be advantageous to have a `Leaving` provider with no when conditions if you want to always know when you no longer have a task.
+
+Secondly, a leaving provider may be executed as a result of it having a causing reaction that another task needs if it has a higher priority than this task.
+For example, if a higher priority kick engine wanted to kick being staticallys stable, and the walk engine had a leaving which caused statically stable, then the leaving .
+In the event that the kick providers task had a lower priority than this providers task, then it is only able to change this module to run other non `Leaving` providers to achive its causing.
+
+In order for a leaving provider to be run however, it must have first been in the entering or normal states.
+The system cannot immediately jump to a leaving provider.
+
+Example:
+
+```cpp
+on<Provides<Walk>>().then()
+```
+
+### When
+
+A when condition is used to ensure that a provider will only execute when its conditions are met.
+These conditions are provided using a global enum message along with a statement that describes what condition you want.
```

Comment:
Syncs everywhere, so hopefully no race conditions in the Director.
Although depending on the exact circumstances it is possible to bounce a state on an off, and confuse the director as it will enable/disable tasks.

But to be honest the director is doing the right thing and the state is just unstable

Reply:
Sounds good to me. Bouncing is just something for the users of it to look out for I think.